pkg/storage/unified/sql/notifier_sql.go

```diff
@@ -0,0 +1,224 @@
+package sql
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/storage/unified/resource"
+	"github.com/grafana/grafana/pkg/storage/unified/sql/sqltemplate"
+	"go.opentelemetry.io/otel/trace"
+)
+
+var (
+	// Validation errors.
+	errHistoryPollRequired    = fmt.Errorf("historyPoll is required")
+	errListLatestRVsRequired  = fmt.Errorf("listLatestRVs is required")
+	errBatchLockRequired      = fmt.Errorf("batchLock is required")
+	errTracerRequired         = fmt.Errorf("tracer is required")
+	errLogRequired            = fmt.Errorf("log is required")
+	errInvalidWatchBufferSize = fmt.Errorf("watchBufferSize must be greater than 0")
+	errInvalidPollingInterval = fmt.Errorf("pollingInterval must be greater than 0")
+	errDoneRequired           = fmt.Errorf("done is required")
+	errDialectRequired        = fmt.Errorf("dialect is required")
+)
+
+// pollingNotifier is a notifier that polls the database for new events.
+type pollingNotifier struct {
+	dialect         sqltemplate.Dialect
+	pollingInterval time.Duration
+	watchBufferSize int
+
+	log    log.Logger
+	tracer trace.Tracer
+
+	batchLock     *batchLock
+	listLatestRVs func(ctx context.Context) (groupResourceRV, error)
+	historyPoll   func(ctx context.Context, grp string, res string, since int64) ([]*historyPollResponse, error)
+
+	done <-chan struct{}
+}
```

Comment:
I extracted all this to break the hard dependency on the backend and make it easier to test.